File: jump.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NAME_SIZE      64
#define PATH_SIZE     256
#define BOOKMARK_SIZE 321

char *bookmarks_filename = "jumprc";

typedef struct bookmark_struct bookmark;
struct bookmark_struct {
    char name[NAME_SIZE];
    char path[PATH_SIZE];
    bookmark *next;
};

bookmark *bookmarks = NULL;

void exit_with_error(char *message) {
    printf("*** ERROR: %s\n", message);
    exit(1);
}

void print_usage() {
    printf("...usage...");
}

char *pwd() {
    char *working_dir;
    FILE *fp;

    working_dir = calloc(PATH_SIZE, sizeof(char));

    fp = popen("pwd", "r");
    if (fp == NULL) {
        exit_with_error("failed to run pwd");
    }
    fgets(working_dir, PATH_SIZE, fp);
    pclose(fp);

    return working_dir;
}

bookmark *add_bookmark(char *name, char *path) {
    bookmark *b;

    // create bookmark node
    b = malloc(sizeof(bookmark));
    strcpy(b->name, name);
    strcpy(b->path, path);
    b->next = NULL;

    // add bookmark to the list
    if (bookmarks == NULL) {
        bookmarks = b;
    } else {
        bookmark *p = bookmarks;

        // find the last but one node
        while (p->next != NULL) {
            p = p->next;
        }

        // add the node to the list
        p->next = b;
    }

    return b;
}

// find a bookmark in the list
bookmark *find_bookmark(char *name) {
    bookmark *p = bookmarks;

    while (p != NULL && strcmp(p->name, name) != 0) {
        p = p->next;
    }

    return p;
}

bookmark *new_bookmark(char *name, char *path) {
    bookmark *p;

    if ((p = find_bookmark(name)) != NULL) {
        strcpy(p->path, path);
    } else {
        p = add_bookmark(name, path);
    }

    return p;
}

void delete_bookmark(char *name) {
    bookmark *p = bookmarks, *d;

    if (strcmp(bookmarks->name, name) == 0) {
        d = bookmarks;
        bookmarks = bookmarks->next;
        free(d);
    } else {
        while (p->next != NULL && strcmp(p->next->name, name) != 0) {
            p = p->next;
        }

        if (p != NULL) {
            d = p->next;
            p->next = d->next;
            free(d);
        }
    }
}

void save_bookmarks() {
    FILE *f;
    bookmark *p = bookmarks;

    if ((f = fopen(bookmarks_filename, "w")) != NULL) {
        while (p != NULL) {
            fprintf(f, "%s\t%s\n", p->name, p->path);
            p = p->next;
        }
    } else {
        exit_with_error("can't open configuration file");
    }
}

void load_bookmarks() {
    FILE *f;
    char name[NAME_SIZE], path[PATH_SIZE];

    if ((f = fopen(bookmarks_filename, "r")) != NULL) {
        while (fscanf(f, "%s\t%s", name, path) != EOF) {
            add_bookmark(name, path);
        }
    } else {
        exit_with_error("can't open configuration file");
    }
}

void print_bookmarks() {
    bookmark *p = bookmarks;

    while (p != NULL) {
        printf("%s\t%s\n", p->name, p->path);
        p = p->next;
    }
}

int main(int argc, char *argv[]) {
    /*
    if (argc == 2) {

    } else { // wrong number of arguments
        print_usage();
        exit(1);
    }
    */

    // load_bookmarks();
    new_bookmark("ciao", "/ciao");
    new_bookmark("ariciao", "/ariciao");
    // save_bookmarks();
    print_bookmarks();
    delete_bookmark("ariciao");
    printf("---\n");
    print_bookmarks();
    return 0;
}

File: jump.rb
#!/usr/bin/env ruby 

# Jump, a bookmarking system for the bash shell.
# Copyright (c) 2010 Giuseppe Capizzi
# mailto: g.capizzi@gmail.com
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

BOOKMARKS_PATH = File.expand_path "~/.jumprc"

def exit_with_error(message)
  puts "*** ERROR: " + message
  exit 1
end

def print_usage
  puts "jump to [bookmark]    Jumps to the directory pointed by [bookmark]"
  puts "jump add [bookmark]   Saves the current directory in [bookmark]"
  puts "jump del [bookmark]   Deletes [bookmark]"
  puts "jump list             Prints the list of all saved bookmarks"
  puts "jump help             Displays this message"
end

# Saves the bookmarks list in the bookmarks file.
def save_bookmarks(bookmarks, bookmarks_path)
  begin
    file = File.open(bookmarks_path, 'w') do |f|
      bookmarks.each do |name, path|
        f.puts name + "\t" + path
      end
    end
  rescue
    exit_with_error "can't save configuration file"
  end
end

# Loads the bookmarks list from the bookmarks file.
def load_bookmarks(bookmarks_path)
  bookmarks = {}
  
  begin
    file = File.open(bookmarks_path, 'r') do |f|
      while line = f.gets  
        name, path = line.split("\t")
        bookmarks[name] = path
      end
    end
    
    return bookmarks
  rescue
    exit_with_error "can't load configuration file"
  end
end

# Prints the bookmarks list.
def print_bookmarks(bookmarks)
  name_col_title = "Bookmark"
  path_col_title = "Path"
  name_col_size = 20
  row_length = 75

  if !bookmarks.empty?
    print name_col_title
    (name_col_size - name_col_title.length).times do print " " end
    puts path_col_title
    puts "-" * row_length

    bookmarks.each do |name, path|
      print name
      (name_col_size - name.length).times do
        print " "
      end
      puts " " + path
    end
  else
    puts "No bookmarks saved"
  end
end

# Expands paths that could start with a bookmark (e.g. [bookmark]/sub/path)
def expand_path(path_with_bookmark, bookmarks)
  if path_with_bookmark.index("/").nil?
    # the path is just a bookmark
    return bookmarks[path_with_bookmark]
  elsif path_with_bookmark.index("/") == 0
    # the path is an absolute path (no bookmark, e.g. /absolute/path)
    return path_with_bookmark
  else
    # the path is composed of a bookmark and a subpath
    name, path = path_with_bookmark.split("/")
    return bookmarks[name] + "/" + path;
  end
end

if __FILE__ == $PROGRAM_NAME
  if ARGV.size > 0
    if ARGV[0] == "list"
      print_bookmarks(load_bookmarks(BOOKMARKS_PATH))
      # p load_bookmarks(BOOKMARKS_PATH).to_s
    elsif ARGV[0] == "help"
      print_usage
    elsif ARGV[0] == "to" && ARGV.size > 1
      path = expand_path(ARGV[1], load_bookmarks(BOOKMARKS_PATH));
      print path.chomp if !path.nil?
    elsif ARGV[0] == "add" && ARGV.size > 1
      bookmarks = load_bookmarks(BOOKMARKS_PATH)
      bookmarks[ARGV[1]] = Dir.pwd
      save_bookmarks(bookmarks, BOOKMARKS_PATH)
    elsif ARGV[0] == "del" && ARGV.size > 1
      bookmarks = load_bookmarks(BOOKMARKS_PATH)
      bookmarks.delete ARGV[1]
      save_bookmarks(bookmarks, BOOKMARKS_PATH)
    else
      print_usage
      exit 1
    end
  else # wrong number of arguments
    print_usage
    exit 1;
  end
end
